dynawo/sources/Modeler/DataInterface/PowSyblIIDM/DYNServiceManagerInterfaceIIDM.cpp

```diff
@@ -0,0 +1,107 @@
+ *
+ */
+
+#include "DYNServiceManagerInterfaceIIDM.h"
+
+#include "DYNDataInterfaceIIDM.h"
+#include "DYNNetworkInterfaceIIDM.h"
+#include "DYNSwitchInterfaceIIDM.h"
+
+namespace DYN {
+
+ServiceManagerInterfaceIIDM::ServiceManagerInterfaceIIDM(const DataInterfaceIIDM* const dataInterface) : dataInterface_(dataInterface), graphs_{} {}
+
+void
+ServiceManagerInterfaceIIDM::buildGraph(const boost::shared_ptr<VoltageLevelInterface>& vl) {
+  auto compare = [](const boost::shared_ptr<BusInterface>& bus, const std::string& id) { return bus->getID() == id; };
+
+  auto& graph = graphs_[vl->getID()];
+  const auto& buses = vl->getBuses();
+
+  for (size_t i = 0; i < buses.size(); i++) {
+    // we are using the position in the vector as the vertex, as these buses are fixed
+    graph.addVertex(i);
+  }
+
+  for (const auto& sw : vl->getSwitches()) {
+    auto busid1 = sw->getBusInterface1()->getID();
+    auto bus1It1 = std::find_if(buses.begin(), buses.end(), std::bind(compare, std::placeholders::_1, busid1));
+#if _DEBUG_
+    // shouldn't happen by construction of the IIDM element
+    assert(bus1It1 != buses.end());
+#endif
+    auto busid2 = sw->getBusInterface2()->getID();
+    auto bus1It2 = std::find_if(buses.begin(), buses.end(), std::bind(compare, std::placeholders::_1, busid2));
+#if _DEBUG_
+    // shouldn't happen by construction of the IIDM element
+    assert(bus1It2 != buses.end());
+#endif
+    // we are using the position of the bus in the bus array as index in the graph, because these indexes won't change during simulation
+    graph.addEdge(bus1It1 - buses.begin(), bus1It2 - buses.begin(), sw->getID());
+  }
+}
+
+std::vector<std::string>
+ServiceManagerInterfaceIIDM::getBusesConnectedBySwitch(const std::string& busId, const std::string& VLId) {
+  const auto& levels = dataInterface_->getNetwork()->getVoltageLevels();
+  auto vlIt = std::find_if(levels.begin(), levels.end(), [&VLId](const boost::shared_ptr<DYN::VoltageLevelInterface>& vl) { return vl->getID() == VLId; });
+  if (vlIt == levels.end()) {
+    throw DYNError(Error::MODELER, UnknownVoltageLevel, VLId);
+  }
+
+  if ((*vlIt)->getVoltageLevelTopologyKind() == VoltageLevelInterface::NODE_BREAKER) {
+    throw DYNError(Error::MODELER, VoltageLevelTopoError, VLId);
+  }
+
+  const auto& buses = (*vlIt)->getBuses();
+  auto it = std::find_if(buses.begin(), buses.end(), [&busId](const boost::shared_ptr<BusInterface>& bus) { return bus->getID() == busId; });
+
+  if (it == buses.end()) {
+    // check next voltage level if bus is not part of it
+    throw DYNError(Error::MODELER, UnknownBus, busId);
+  }
+
+  if (graphs_.count((*vlIt)->getID()) == 0) {
+    buildGraph(*vlIt);
+  }
+
+  // Change weight of edges
+  boost::unordered_map<std::string, float> weights;
+  for (const auto& sw : (*vlIt)->getSwitches()) {
+    weights[sw->getID()] = sw->isOpen() ? 0 : 1;
+  }
+
+  std::vector<std::string> ret;
+
+  size_t busIndexFound = it - buses.begin();
```

Comment:
why using an iterator instead of directly comparing the id within the loop?

Reply:
Dyn Graph is using indexes so we use indexes as well to avoid computing the distance at each step of the loop

Reply:
yes but we go though the buses just below so we have a direct access to their ids

Reply:
ids, yes not their index

Reply:
Let me rephrase: buses contain the BusInterface. Above you search which iterator corresponds to bus->getID() == busId, and then you compare this index to busIndex.
Why not replacing all of this by if (buses[busIndex]->getID() == busId) that would be way more readable?